### DISCUSSION TITLE
Difference between Until Executed and Until And While Executing

### DISCUSSION THREAD

**Discussion by LucDelmon** [mhenrixon/sidekiq-unique-jobs#660] [Q&A] [answered]
Hi,

Sorry if it might seems like a trivial question, but after reading the doc severals times I have the feeling that "Until Executed" and "Until And While Executing" locking strategy do the exact same thing.

From what I understand "Until Executed" will put a lock when the job is enqueued and remove it after `#perform` execution (in a server middleware I guess).

And "Until And While Executing" will put a lock when the job is enqueued remove it when job `#perform` is called but immediately put it again during the execution.

So what is the difference in the end. Are the lock during "Until And While Executing" 2 different locks ?

Or is the difference is in the way they handle retry ? I can't get around it.

If it helps, what I'm looking for is the strategy that will put a lock from first enqueing to '#perform' success without removing the lock during retries or at any time.

I guess it's one of them but I'm not sure which one.

Thanks in advance

**Comment by LucDelmon** [comment, marked as answer]
looking at other discussion I found this answer in a thread https://github.com/mhenrixon/sidekiq-unique-jobs/discussions/656#discussioncomment-1646774 
It seems that the difference is indeed that the 2 locks are different. Maybe the doc for "Until And While Executing" could be more clear and specify that the combination of the 2 locks is really an addition of 2 different lock sand not a combination of behavior in 1 lock.

I'm still not quite sure of the behavior in retries but I should be able to try it myself.

**Reply by mhenrixon** [reply]
Documentation could certainly be improved. I have been considering splitting it up in separate markdown files to give a better overview. 

The readme is quite long already.
